test/e2e/util/oauth.go

```diff
@@ -0,0 +1,208 @@
+package util
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/url"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	v1 "github.com/openshift/api/config/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+	hcpmanifests "github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/manifests"
+	configmanifests "github.com/openshift/hypershift/control-plane-operator/hostedclusterconfigoperator/controllers/resources/manifests"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/clientcmd"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	userv1 "github.com/openshift/api/user/v1"
+	userv1client "github.com/openshift/client-go/user/clientset/versioned/typed/user/v1"
+
+	restclient "k8s.io/client-go/rest"
+)
+
+func TestHtpasswdIdentityProvider(t *testing.T, ctx context.Context, client crclient.Client, hostedCluster *hyperv1.HostedCluster) {
+	t.Run("TestHtpasswdIdentityProvider", func(t *testing.T) {
+		g := NewWithT(t)
+
+		// secret containing htpasswd "file": `htpasswd -cbB htpasswd.tmp testuser password`
+		secret := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "htpasswd",
+				Namespace: hostedCluster.Namespace},
+			Data: map[string][]byte{
+				"htpasswd": []byte("testuser:$2y$05$0Fk2s.0FbLy0FZ82JAqajOV/kbT/wqKX5/QFKgps6J69J2jY6r5ZG"),
+			},
+		}
+
+		err := client.Create(ctx, &secret)
+		g.Expect(err).ToNot(HaveOccurred(), "failed to create htpasswd secret")
+
+		if hostedCluster.Spec.Configuration == nil {
+			hostedCluster.Spec.Configuration = &hyperv1.ClusterConfiguration{}
+		}
+		hostedCluster.Spec.Configuration.OAuth = &v1.OAuthSpec{
+			IdentityProviders: []v1.IdentityProvider{
+				{
+					Name:          "my_htpasswd_provider",
+					MappingMethod: v1.MappingMethodClaim,
+					IdentityProviderConfig: v1.IdentityProviderConfig{
+						Type: v1.IdentityProviderTypeHTPasswd,
+						HTPasswd: &v1.HTPasswdIdentityProvider{
+							FileData: v1.SecretNameReference{
+								Name: secret.Name,
+							},
+						},
+					},
+				},
+			},
+		}
+
+		err = client.Update(ctx, hostedCluster)
+		g.Expect(err).ToNot(HaveOccurred(), "failed to update hostedcluster identity providers")
+
+		guestKubeConfigSecretData, err := WaitForGuestKubeConfig(t, ctx, client, hostedCluster)
```

Comment:
why doing this instead of `guestClient := e2eutil.WaitForGuestClient(t, testContext, client, hostedCluster)`?

Reply:
I only need the `restclient.Config` to create the `transport` used for the http client. `WaitForGuestClient` would return a `crclient.Client` that doesn't have a rest config that I can use.